homeassistant/components/met/weather.py

```diff
@@ -62,101 +58,80 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
     if config.get(CONF_LATITUDE) is None:
         config[CONF_TRACK_HOME] = True
 
-    async_add_entities([MetWeather(config, hass.config.units.is_metric)])
+    await async_setup(hass, config, async_add_entities)
 
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
     """Add a weather entity from a config_entry."""
-    async_add_entities([MetWeather(config_entry.data, hass.config.units.is_metric)])
+    await async_setup(hass, config_entry.data, async_add_entities)
+
+
+async def async_setup(hass, config, async_add_entities):
+    """Create DataUpdateCoordinator and add entities."""
+    if config.get(CONF_TRACK_HOME, False):
+        unique_id = "home"
+    else:
+        unique_id = f"{config[CONF_LATITUDE]}-{config[CONF_LONGITUDE]}"
+
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            return await hass.data[DOMAIN][unique_id].fetch_data()
+        except Exception as err:
+            raise UpdateFailed(f"Update failed: {err}")
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=unique_id,
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=timedelta(minutes=randrange(55, 65)),
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    hass.data[DOMAIN][unique_id].init_data()
+    await hass.data[DOMAIN][unique_id].fetch_data()
+
+    async_add_entities([MetWeather(coordinator, config, hass.config.units.is_metric)])
 
 
 class MetWeather(WeatherEntity):
     """Implementation of a Met.no weather condition."""
 
-    def __init__(self, config, is_metric):
+    def __init__(self, coordinator, config, is_metric):
         """Initialise the platform with a data instance and site."""
         self._config = config
+        self._coordinator = coordinator
         self._is_metric = is_metric
         self._unsub_track_home = None
-        self._unsub_fetch_data = None
-        self._weather_data = None
-        self._current_weather_data = {}
-        self._coordinates = {}
-        self._forecast_data = None
 
     async def async_added_to_hass(self):
         """Start fetching data."""
-        await self._init_data()
+        self.async_on_remove(
+            self._coordinator.async_add_listener(self.async_write_ha_state)
+        )
         if self._config.get(CONF_TRACK_HOME):
             self._unsub_track_home = self.hass.bus.async_listen(
-                EVENT_CORE_CONFIG_UPDATE, self._init_data
+                EVENT_CORE_CONFIG_UPDATE, self._core_config_updated
             )
+        await self.async_update()
```

Comment:
Here the data is also fetched again during setup, but through the coordinator this time.

Reply:
Aaa, got it. Thanks. I removed fetching the data after init and it works.  And all tests have passed now! Thanks for your help and patience!